### PR TITLE
fix: TypeError when using generic types in resolver in more complex c…

### PR DESCRIPTION
**Pull request by invokermain**
…ases

## Description

Fixes a field's resolvers from raising a type error when using a subclassed generic return type hint by widening the check for generics in the `StrawberryField.type` property, [original code here](https://github.com/strawberry-graphql/strawberry/blob/main/strawberry/field.py#L200).

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/strawberry-graphql/strawberry/issues/1899

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Tim OSullivan for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by invokermain** [comment]
Holding off on adding a RELEASE.md file as would like some discussion around this first:

1. Is it a good idea to use `StrawberryType.is_generic` for this?
2. Is `tests/schema/test_resolvers.py` the best location for the tests? (slightly overwhelmed by the number of unit tests!)
3. Follow on for above, is there a more granular/small unit that could be used to tests these, my understanding of the internals of this library is too poor to know the answer to this. Maybe these should be unit tests on `GraphQLCoreConverter` instead?
4. Are there more edge cases we should test against? Maybe `Union[T, SomeThingElse]`, for example if someone was returning error responses as part of their schema or the generic type.
5. The big question (probably should have been asked before the previous pull request), if we are evading the `MissingTypesForGenericError` so much, I wonder what else uses it to make it useful? If the answer is not much, then I wonder what happens if we don't raise that error.

Ultimately I am guessing these PRs are a short term solution to allow you to type hint resolver factories correctly, and in the long term you would like to maybe check that say `Container[SolidType]` & `Container[GenericType]` are compatible. Although that sounds complicated! But that would be part of the warning/incorrect type checking logic hinted at in: https://github.com/strawberry-graphql/strawberry/issues/1896

**Comment by patrick91** [comment]
> Holding off on adding a RELEASE.md file as would like some discussion around this first:
> 
> 1. Is it a good idea to use `StrawberryType.is_generic` for this?

Based on our current code structure, I'd say yes. I was a bit worried that doing that would break some code, but we do have quite a bit of tests for generics so we should be fine :)

> 2. Is `tests/schema/test_resolvers.py` the best location for the tests? (slightly overwhelmed by the number of unit tests!)

I think so, we don't have a great structure for tests at the moment, so I wouldn't worry too much 😊

> 3. Follow on for above, is there a more granular/small unit that could be used to tests these, my understanding of the internals of this library is too poor to know the answer to this. Maybe these should be unit tests on `GraphQLCoreConverter` instead?

The test you wrote is fine for me, writing a more granular test might be good, but we still want to make sure the resolvers works. Maybe we can add a test for the schema, to check that the types are what we expect :)

> 4. Are there more edge cases we should test against? Maybe `Union[T, SomeThingElse]`, for example if someone was returning error responses as part of their schema or the generic type.

Union[T, SomethingElse], Optional[T] and List[T], not sure if we should test more cases

> 5. The big question (probably should have been asked before the previous pull request), if we are evading the `MissingTypesForGenericError` so much, I wonder what else uses it to make it useful? If the answer is not much, then I wonder what happens if we don't raise that error.

I'm going to review all our expected errors at some point, so I wouldn't change this too much for now 😊

> Ultimately I am guessing these PRs are a short term solution to allow you to type hint resolver factories correctly, and in the long term you would like to maybe check that say `Container[SolidType]` & `Container[GenericType]` are compatible. Although that sounds complicated! But that would be part of the warning/incorrect type checking logic hinted at in: #1896

I'd worry about this in future, we can definitely add some basic checks which might cover most of the common errors 😊 we'll see 😊

**Comment by invokermain** [comment]
I have added some extra unit tests covering the Union[T, SomethingElse], Optional[T] and List[T] cases.

> Maybe we can add a test for the schema, to check that the types are what we expect :)
With regards to this happy to add them with more pointers. Currently I am not 100% sure if its required/what is required as not sure if any of my code changes affect the schema generation?

I was under the impression that the Schema is generated from the top level type hint (e.g. [here ](https://gist.github.com/invokermain/31c783c76dd26e1da862cbe5218d2500#file-strawb_generic_example-py-L51) it would be defined by the `list[MyItemType]` type hint no?).

**Comment by invokermain** [comment]
playground for example in RELEASE.md: https://play.strawberry.rocks/?gist=715c09744893b8662b8468f29579ee90

File: strawberry/field.py
Comment (patrick91) [inline strawberry/field.py:59]: adding a note for the future (maybe), we could move this method on the resolver itself potentially

**Review by patrick91 — APPROVED** [review]
Apologies for the very late review!

I've merged main into this PR and reviewed the code, looks good! I've left a minor note for the future, but I'll merge this now when tests are green :)

**Comment by invokermain** [comment]
> Apologies for the very late review!
> 
> I've merged main into this PR and reviewed the code, looks good! I've left a minor note for the future, but I'll merge this now when tests are green :)

No worries! It was a very minor issue, but just means I can uncomment some generic type hints now, thanks 👍

**Pull request merged**
